finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/RawZipkinTracer.scala

```diff
@@ -0,0 +1,149 @@
+      case tracing.Annotation.BinaryAnnotation(key: String, value: Boolean) =>
+        binaryAnnotation(record, key, (if (value) TrueBB else FalseBB).duplicate(), thrift.AnnotationType.BOOL)
+      case tracing.Annotation.BinaryAnnotation(key: String, value: Array[Byte]) =>
+        binaryAnnotation(record, key, ByteBuffer.wrap(value), thrift.AnnotationType.BYTES)
+      case tracing.Annotation.BinaryAnnotation(key: String, value: ByteBuffer) =>
+        binaryAnnotation(record, key, value, thrift.AnnotationType.BYTES)
+      case tracing.Annotation.BinaryAnnotation(key: String, value: Short) =>
+        binaryAnnotation(record, key, ByteBuffer.allocate(2).putShort(0, value), thrift.AnnotationType.I16)
+      case tracing.Annotation.BinaryAnnotation(key: String, value: Int) =>
+        binaryAnnotation(record, key, ByteBuffer.allocate(4).putInt(0, value), thrift.AnnotationType.I32)
+      case tracing.Annotation.BinaryAnnotation(key: String, value: Long) =>
+        binaryAnnotation(record, key, ByteBuffer.allocate(8).putLong(0, value), thrift.AnnotationType.I64)
+      case tracing.Annotation.BinaryAnnotation(key: String, value: Double) =>
+        binaryAnnotation(record, key, ByteBuffer.allocate(8).putDouble(0, value), thrift.AnnotationType.DOUBLE)
+      case tracing.Annotation.BinaryAnnotation(key: String, value: String) =>
+        binaryAnnotation(record, key, ByteBuffer.wrap(value.getBytes), thrift.AnnotationType.STRING)
+      case tracing.Annotation.BinaryAnnotation(key: String, value) => // Throw error?
+      case tracing.Annotation.LocalAddr(ia: InetSocketAddress) =>
+        setEndpoint(record, ia)
+      case tracing.Annotation.ClientAddr(ia: InetSocketAddress) =>
+        // use a binary annotation over a regular annotation to avoid a misleading timestamp
+        spanMap.update(record.traceId) { _.addBinaryAnnotation(BinaryAnnotation(
+          thrift.Constants.CLIENT_ADDR,
+          TrueBB.duplicate(),
+          thrift.AnnotationType.BOOL,
+          Endpoint.fromSocketAddress(ia)))
+        }
+      case tracing.Annotation.ServerAddr(ia: InetSocketAddress) =>
+        spanMap.update(record.traceId) { _.addBinaryAnnotation(BinaryAnnotation(
+          thrift.Constants.SERVER_ADDR,
+          TrueBB.duplicate(),
+          thrift.AnnotationType.BOOL,
+          Endpoint.fromSocketAddress(ia)))
+        }
+    }
+  }
+
+  /**
+   * Sets the endpoint in the span for any future annotations. Also
+   * sets the endpoint in any previous annotations that lack one.
+   */
+  protected def setEndpoint(record: Record, ia: InetSocketAddress) {
+    spanMap.update(record.traceId)(_.setEndpoint(Endpoint.fromSocketAddress(ia).boundEndpoint))
+  }
+
+  protected def binaryAnnotation(
+    record: Record,
+    key: String,
+    value: ByteBuffer,
+    annotationType: thrift.AnnotationType
+  ) {
+    spanMap.update(record.traceId) { span =>
+      span.addBinaryAnnotation(BinaryAnnotation(key, value, annotationType, span.endpoint))
```

Comment:
ex in my local workspace, I made this a map of zipkin.Span.Builder.. not required, as such a change is internal-only and could be layered as easily later